### PR TITLE
[FEATURE] Implement MulticolumnMapExpectation class

### PR DESCRIPTION
**Pull request by alexsherstinsky**
Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html).

Changes proposed in this pull request:
- JIRA: GE-320/GE-401
-
-

After submitting your PR, CI checks will run and @tiny-tim-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/en/latest/contributing/style_guide.html?highlight=style%20guide)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html?highlight=checklist) of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added [unit tests](https://docs.greatexpectations.io/en/latest/contributing/testing.html#contributing-testing-writing-unit-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

**Comment by alexsherstinsky** [comment]
### IMPORTANT
This PR requires PR #3130 to be merged first as a prerequisite.

File: great_expectations/expectations/expectation.py [hunk 1]
```diff
@@ -1367,6 +1367,188 @@ def _validate(
...
+            configuration=configuration,
+            runtime_configuration=runtime_configuration,
+        )
+        metric_dependencies[
+            self.map_metric + ".filtered_row_count"
+        ] = MetricConfiguration(
+            metric_name=self.map_metric + ".filtered_row_count",
+            metric_domain_kwargs=metric_kwargs["metric_domain_kwargs"],
+            metric_value_kwargs=metric_kwargs["metric_value_kwargs"],
+        )
```
Comment (jcampbell) [inline great_expectations/expectations/expectation.py:1459]: I think this approach is *much* more readable. Like, *so* much more readable than the way the `ColumnMapExpectation` handles this that I think we should add "filtered_row_count" to all `MapMetricProvider` children. If you agree, let's add it to the JIRA regarding streamlining the code that handles the metric registration process.

File: great_expectations/expectations/expectation.py [hunk 2]
```diff
@@ -1367,6 +1367,188 @@ def _validate(
...
+            )
+
+        total_count = metrics.get("table.row_count")
+        unexpected_count = metrics.get(self.map_metric + ".unexpected_count")
+        unexpected_values = metrics.get(self.map_metric + ".unexpected_values")
+        unexpected_index_list = metrics.get(self.map_metric + ".unexpected_index_list")
+        filtered_row_count = metrics.get(self.map_metric + ".filtered_row_count")
+
+        success = None
+        if total_count is None:
```
Comment (jcampbell) [inline great_expectations/expectations/expectation.py]: Should we add this condition?

```suggestion
        if total_count is None or filtered_row_count is None or total_count == 0 or filtered_row_count == 0:
```

File: great_expectations/expectations/expectation.py [hunk 3]
```diff
@@ -1367,6 +1367,188 @@ def _validate(
...
+            # Vacuously true
+            success = True
+        elif total_count != 0:
+            success = unexpected_count == 0
```
Comment (jcampbell) [inline great_expectations/expectations/expectation.py:1541]: See above re: support for `mostly`

File: great_expectations/expectations/expectation.py [hunk 4]
```diff
@@ -1367,6 +1367,188 @@ def _validate(
...
+class MulticolumnMapExpectation(TableExpectation, ABC):
+    map_metric = None
+    domain_keys = (
+        "batch_id",
+        "table",
+        "column_list",
+        "row_condition",
+        "condition_parser",
+    )
+    success_keys = tuple()
```
Comment (jcampbell) [inline great_expectations/expectations/expectation.py:1379]: Shouldn't we support mostly?

**Review by jcampbell — CHANGES_REQUESTED** [review]

File: great_expectations/expectations/expectation.py [hunk 5]
```diff
@@ -1367,6 +1367,188 @@ def _validate(
         )
 
 
+class MulticolumnMapExpectation(TableExpectation, ABC):
```
Comment (jcampbell) [inline great_expectations/expectations/expectation.py:1370]: Please revise capitalization per: result of slack convo inside the team

File: great_expectations/expectations/expectation.py (hunk 1 above)
Comment (alexsherstinsky) [reply great_expectations/expectations/expectation.py:1459]: @jcampbell Agreed.  This is now in JIRA: `https://superconductive.atlassian.net/browse/GE-439`.

File: great_expectations/expectations/expectation.py (hunk 2 above)
Comment (alexsherstinsky) [reply great_expectations/expectations/expectation.py]: @jcampbell I agree with the idea -- my suggestion is to make it read as follows:
```
        if total_count is None or filtered_row_count is None or total_count == 0 or filtered_row_count == 0:
            # Vacuously true
            success = True
        else:
            success = unexpected_count == 0
```
Please let me know what you think.  Thanks.

File: great_expectations/expectations/expectation.py (hunk 4 above)
Comment (alexsherstinsky) [reply great_expectations/expectations/expectation.py:1379]: @jcampbell In my opinion, `mostly` would belong in a separate PR, and I think that `mostly` does make sense for multicolumn expectations.

File: great_expectations/expectations/expectation.py (hunk 3 above)
Comment (alexsherstinsky) [reply great_expectations/expectations/expectation.py:1541]: @jcampbell In my opinion, `mostly` would belong in a separate PR, and I think that `mostly` does make sense for multicolumn expectations.  If agree, I will add a JIRA task for it.  Thanks.

File: great_expectations/expectations/expectation.py (hunk 5 above)
Comment (alexsherstinsky) [reply great_expectations/expectations/expectation.py:1370]: @jcampbell I cannot tell with certainty what the consensus in Slack is for the naming.  It seems that the Team prefer `MultiColumn` -- no problem.  How about we update this uniformly across the expectations and metrics as quick follow-on PR?  If agree, I will add it to JIRA and take care of it right away.  Thanks.

Comment (jcampbell) [reply great_expectations/expectations/expectation.py:1370]: Sounds good.

File: great_expectations/expectations/expectation.py (hunk 4 above)
Comment (jcampbell) [reply great_expectations/expectations/expectation.py:1379]: Works for me.

File: great_expectations/expectations/expectation.py (hunk 2 above)
Comment (jcampbell) [reply great_expectations/expectations/expectation.py]: +1

File: great_expectations/expectations/expectation.py (hunk 3 above)
Comment (jcampbell) [reply great_expectations/expectations/expectation.py:1541]: Yes please on JIRA!

File: great_expectations/expectations/expectation.py (hunk 2 above)
Comment (jcampbell) [reply great_expectations/expectations/expectation.py]: Updated the comment and will commit suggestion to be able to mark as approved.

**Review by jcampbell — APPROVED** [review]

File: great_expectations/expectations/expectation.py (hunk 3 above)
Comment (alexsherstinsky) [reply great_expectations/expectations/expectation.py:1541]: @jcampbell The relevant new JIRA items are: `https://superconductive.atlassian.net/browse/GE-440` and `https://superconductive.atlassian.net/browse/GE-441` (unassigned for now; anybody can do them).  Thanks.

File: great_expectations/expectations/expectation.py (hunk 5 above)
Comment (alexsherstinsky) [reply great_expectations/expectations/expectation.py:1370]: @jcampbell -- This is JIRA `https://superconductive.atlassian.net/browse/GE-442` (and it contains a couple of quick questions for you).  Thanks.

**Pull request merged**